### PR TITLE
Added code review recipes for Terraform

### PR DESCRIPTION
**Pull request by ritesh-modi**
Added few code review recipes related to Terraform in this request. Please let me know feedback/comments on the same.

### CODE REVIEW

**Review by codepossible — APPROVED** [review]
LGTM

File: code-reviews/recipes/Terraform.md [hunk 1]
```diff
@@ -86,3 +86,13 @@ In addition to the [Code Review Checklist](../readme.md) you should also look fo
 * [ ] Is `try` function used only with simple attribute references and type conversion functions?, as overuse of `try` function to suppress errors will lead to a configuration that is hard to understand and maintain
 * [ ] Are the explicit type conversion functions used to normalize types returned only in module outputs?, as the explicit type conversions are rarely necessary in Terraform because it will convert types automatically where required
 * [ ] Is `Sensitive` property on schema set to `true` for the fields that contains sensitive information?. This will prevent the field's values from showing up in CLI output
+* [ ] Each defined variables should explicitly have `type information`. E.g. a list(string) or string
```
Comment (davidkpiano) [inline code-reviews/recipes/Terraform.md:89]:
```suggestion
* [ ] Each defined variables should explicitly have type information. E.g. a `list(string)` or `string`
```

File: code-reviews/recipes/Terraform.md [hunk 2]
```diff
@@ -86,3 +86,13 @@ In addition to the [Code Review Checklist](../readme.md) you should also look fo
...
+* [ ] Each defined variables should explicitly have `description` about the purpose of variable and usage
+* [ ] Don’t provide a `default value` for a variable the value of which must be supplied by a user.
```
Comment (davidkpiano) [inline code-reviews/recipes/Terraform.md:91]:
```suggestion
* [ ] Don’t provide a `default` value for a variable the value of which must be supplied by a user.
```

File: code-reviews/recipes/Terraform.md [hunk 3]
```diff
@@ -86,3 +86,13 @@ In addition to the [Code Review Checklist](../readme.md) you should also look fo
...
+* [ ] Try avoiding nesting sub configuration within resources. Create a sepearate resource section for resources even though they can be decalred as sub-element of a resource. For example, declaring subnets within virtual network vis-a-vis declaring subnets as a separate resources compared to virtual network on Azure
+* [ ] Never `hard code` any value in configuration. Declare them from locals section if needed multiple times as static value and are internal to the configuration.
```
Comment (davidkpiano) [inline code-reviews/recipes/Terraform.md:93]:
```suggestion
* [ ] Never hard-code any value in configuration. Declare them from locals section if needed multiple times as static value and are internal to the configuration.
```

File: code-reviews/recipes/Terraform.md [hunk 4]
```diff
@@ -86,3 +86,13 @@ In addition to the [Code Review Checklist](../readme.md) you should also look fo
...
+* [ ] The `names of the resources` created on Azure should not be hard-coded or static. These names should be dynamic and user provided using variables. This is helpful especially in unit testing when multiple tests are running in parallel trying to create resources on Azure but need different names ( few names in azure needs to be named uniquely e.g. storage accounts).
```
Comment (davidkpiano) [inline code-reviews/recipes/Terraform.md:94]:
```suggestion
* [ ] The `name`s of the resources created on Azure should not be hard-coded or static. These names should be dynamic and user-provided using `variable` blocks. This is helpful especially in unit testing when multiple tests are running in parallel trying to create resources on Azure but need different names (few names in Azure need to be named uniquely; e.g. storage accounts).
```

File: code-reviews/recipes/Terraform.md [hunk 5]
```diff
@@ -86,3 +86,13 @@ In addition to the [Code Review Checklist](../readme.md) you should also look fo
...
+* [ ] It is a good practice to `output` the `ID of resources` created on Azure from configuration. This is especially helpful when adding dynamic blocks for sub-elements/child elements to the parent resource.
```
Comment (davidkpiano) [inline code-reviews/recipes/Terraform.md:95]:
```suggestion
* [ ] It is a good practice to `output` the ID of resources created on Azure from configuration. This is especially helpful when adding dynamic blocks for sub-elements/child elements to the parent resource.
```

File: code-reviews/recipes/Terraform.md [hunk 6]
```diff
@@ -86,3 +86,13 @@ In addition to the [Code Review Checklist](../readme.md) you should also look fo
...
+* [ ] Use `required_providers` block for establishing the dependency for providers along with pre-determined version.
+* [ ] Use `terraform block` to declare the provider dependency with exact version and also the terraform CLI version needed for the configuration
```
Comment (davidkpiano) [inline code-reviews/recipes/Terraform.md:97]:
```suggestion
* [ ] Use `terraform` block to declare the provider dependency with exact version and also the terraform CLI version needed for the configuration
```

File: code-reviews/recipes/Terraform.md [hunk 7]
```diff
@@ -86,3 +86,13 @@ In addition to the [Code Review Checklist](../readme.md) you should also look fo
...
+* [ ] Validate the variables values supplied based on usage and type of variable. The validation can be done to variables by adding `validation block`.
```
Comment (davidkpiano) [inline code-reviews/recipes/Terraform.md:98]:
```suggestion
* [ ] Validate the variables values supplied based on usage and type of variable. The validation can be done to variables by adding `validation` block.
```

**Review by davidkpiano — COMMENTED** [review]
Adding a few suggestions to fix code formatting/grammar

File: code-reviews/recipes/Terraform.md (hunk 1 above)
Comment (ritesh-modi) [reply code-reviews/recipes/Terraform.md:89]: Great point.. Change accepted

File: code-reviews/recipes/Terraform.md (hunk 2 above)
Comment (ritesh-modi) [reply code-reviews/recipes/Terraform.md:91]: Great point.. Change accepted

File: code-reviews/recipes/Terraform.md (hunk 3 above)
Comment (ritesh-modi) [reply code-reviews/recipes/Terraform.md:93]: Great point.. Change accepted

File: code-reviews/recipes/Terraform.md (hunk 5 above)
Comment (ritesh-modi) [reply code-reviews/recipes/Terraform.md:95]: Great point.. Change accepted

File: code-reviews/recipes/Terraform.md (hunk 6 above)
Comment (ritesh-modi) [reply code-reviews/recipes/Terraform.md:97]: Great point.. Change accepted

File: code-reviews/recipes/Terraform.md [hunk 8]
```diff
@@ -86,3 +86,13 @@ In addition to the [Code Review Checklist](../readme.md) you should also look fo
 * [ ] Is `try` function used only with simple attribute references and type conversion functions?, as overuse of `try` function to suppress errors will lead to a configuration that is hard to understand and maintain
 * [ ] Are the explicit type conversion functions used to normalize types returned only in module outputs?, as the explicit type conversions are rarely necessary in Terraform because it will convert types automatically where required
 * [ ] Is `Sensitive` property on schema set to `true` for the fields that contains sensitive information?. This will prevent the field's values from showing up in CLI output
+* [ ] Each defined variables should explicitly have `type` information. E.g. a `list(string)` or `string`
+* [ ] Each defined variables should explicitly have `description` about the purpose of variable and usage
+* [ ] Don’t provide a `default` value for a variable the value of which must be supplied by a user.
```
Comment (p-schuler) [inline code-reviews/recipes/Terraform.md:91]: nit: simplify?

 Don’t provide a `default` value for a variable which must be supplied by a user.

Comment (p-schuler) [reply code-reviews/recipes/Terraform.md:91]: Also, can we make sure we are consistent using an ending '.' at each list item? Existing list does not have a dot at the end, your items are inconsistent in use.

File: code-reviews/recipes/Terraform.md [hunk 9]
```diff
@@ -86,3 +86,13 @@ In addition to the [Code Review Checklist](../readme.md) you should also look fo
...
+* [ ] Never hard-code any value in configuration. Declare them from `locals` section if a variable is needed multiple times as a static value and are internal to the configuration.
+* [ ] The `name`s of the resources created on Azure should not be hard-coded or static. These names should be dynamic and user-provided using `variable` blocks. This is helpful especially in unit testing when multiple tests are running in parallel trying to create resources on Azure but need different names (few names in Azure need to be named uniquely; e.g. storage accounts).
+* [ ] It is a good practice to `output` the ID of resources created on Azure from configuration. This is especially helpful when adding dynamic blocks for sub-elements/child elements to the parent resource.
+* [ ] Use `required_providers` block for establishing the dependency for providers along with pre-determined version.
+* [ ] Use `terraform` block to declare the provider dependency with exact version and also the terraform CLI version needed for the configuration
+* [ ] Validate the variables values supplied based on usage and type of variable. The validation can be done to variables by adding `validation` block.
```
Comment (p-schuler) [inline code-reviews/recipes/Terraform.md:98]: variable's values?

**Review by p-schuler — APPROVED** [review]

File: code-reviews/recipes/Terraform.md (hunk 9 above)
Comment (RIteshSolidity) [reply code-reviews/recipes/Terraform.md]: feedback accepted

File: code-reviews/recipes/Terraform.md (hunk 8 above)
Comment (ritesh-modi) [reply code-reviews/recipes/Terraform.md]: feedback accepted

File: code-reviews/recipes/Terraform.md (hunk 9 above)
Comment (ritesh-modi) [reply code-reviews/recipes/Terraform.md]: feedback accepted

File: code-reviews/recipes/Terraform.md (hunk 8 above)
Comment (ritesh-modi) [reply code-reviews/recipes/Terraform.md]: and fixed

File: code-reviews/recipes/Terraform.md (hunk 9 above)
Comment (ritesh-modi) [reply code-reviews/recipes/Terraform.md]: and fixed

**Comment by ritesh-modi** [comment]
all changes incorporated based on feedbacks

**Comment by fnocera** [comment]
@ritesh-modi please feel free to complete if you have addressed all the comments!

**Comment by davidkpiano** [comment]
@ritesh-modi Did you mean to close this?

**Pull request merged**